### DISCUSSION TITLE
validParams check on FlowConvection Kernel

### DISCUSSION THREAD

**Discussion by ghost** [shannon-lab/zapdos#112] [Q&A Plasma Transport] [answered]
Hello everyone

Since I updated the MOOSE suite a couple days ago, I've noticed a warning displayed on my terminal each time I run something due to my custom FlowConvection kernel: 

```
*** Warning, This code is deprecated and will be removed in future versions:
The following object(s) are constructed using the legacy input parameter construction:

  FlowConvection (ZapdosApp)

Convert InputParameters validParams<T>() for each object into a static
member function InputParameters T::validParams() and remove the old function.
```

So I tried my hand at fixing the problem, hopefully just by following the instructions. I attached the a txt file with both the .C and .h code, and I made what I deemed to be the necessary edits and left comments by them. If anyone has a few spare minutes and can quickly check my work, I would greatly appreciate it!

Edit: I tried to compile as is and I got the following errors:

```
/home/sean_wsl/projects/zapdos/build/header_symlinks/FlowConvection.h:18:45: error: invalid use of incomplete type 'class FlowConvection'
   18 | InputParameters FlowConvection::validParams( ); //* Used to be InputParameters validParams<FlowConvection>();
      |                                                                              ^
/home/sean_wsl/projects/zapdos/build/header_symlinks/FlowConvection.h:15:7: note: forward declaration of 'class FlowConvection'
   15 | class FlowConvection;
      |           ^~~~~~~~~~
In file included from /home/sean_wsl/projects/zapdos/build/unity_src/kernels_Unity.C:21:
/home/sean_wsl/projects/zapdos/src/kernels/FlowConvection.C:21:29: error: no member function 'validParams' declared in
FlowConvection'
   21 | FlowConvection::validParams( ) //* Used to be validParams<FlowConvection>()
      |                                                  ^
```

[FlowConvection_src.txt](https://github.com/shannon-lab/zapdos/files/8150361/FlowConvection_src.txt)

**Comment by cticenhour** [comment, marked as answer]
Just looking at the error output, it appears as though it's not interpreting your comments correctly. I would remove them and try to build again. 

This warning is just saying that the old (circa two years ago and before) method of declaring valid object parameters is being removed from the framework....which is actually happening very soon (see idaholab/moose#19440). If you would like to make a pull request fixing this object up, we would welcome it! 

To make looking at code changes easier, I would highly recommend that you create a git branch and commit your code changes. This lets us look at a summary of your changes and make direct suggestions in the code. Look for some tips on how to work in git that I made for Zapdos here starting with the heading "Work in a GitHub Fork": https://shannon-lab.github.io/zapdos/development/contributing.html

**Reply by cticenhour** [reply]
OK - I've now got the chance to look at the source code. You need to make a small adjustment to your header. Where you have:

```
class FlowConvection;

template <>
InputParameters FlowConvection::validParams(); //* Used to be InputParameters validParams<FlowConvection>();
```

should be removed, as you don't need the template any longer (which declared the previous version of validParams) and neither do you need the forward class declaration as the template is no more. 

Now we need to properly declare the class version of validParams. It's a public function (we want anything inheriting from this class to have access to it), so it'll go in the public section of the class declaration. Further, it's a C++ static function, so we need to add this full line:

```
static InputParameters validParams();
```
 
In your source code (`.C`) file, you also need to remove the line `template <>`, as validParams is no longer a template function.
